packages/react/src/authentication/useSession.ts

```diff
@@ -121,38 +141,167 @@ export type Session = AnonymousSession | ProfileSession | WalletOnlySession;
  * }
  * ```
  *
+ * ## Suspense Enabled
+ *
+ * You can enable suspense mode to suspend the component until the session data is available.
+ *
+ * ```tsx
+ * function Page() {
+ *   const { data } = useSession({ suspense: true });
+ *
+ *   switch (data.type) {
+ *     case SessionType.Anonymous:
+ *       // data is a AnonymousSession
+ *       return <Login />;
+ *
+ *     case SessionType.JustWallet:
+ *       // data is a WalletOnlySession
+ *       return <MyWallet address={data.address} />;
+ *
+ *     case SessionType.WithProfile:
+ *       // data is a ProfileSession
+ *       return <MyProfile profile={data.profile} />;
+ *
+ *     default:
+ *       return <p>Something went wrong.</p>;
+ *   }
+ * }
+ * ```
+ *
+ * Further session data updates will NOT trigger a suspense.
+ *
  * @category Authentication
  * @group Hooks
  */
-export function useSession(): ReadResult<Session, UnspecifiedError> {
+export function useSession(args: UseSessionArgs<true>): SuspenseReadResult<Session>;
+export function useSession(args?: UseSessionArgs<never>): ReadResult<Session, UnspecifiedError>;
+export function useSession(
+  args?: UseSessionArgs<boolean>,
+): ReadResult<Session, UnspecifiedError> | SuspenseReadResult<Session> {
   const sessionData = useSessionDataVar();
 
-  const prevSession = usePreviousValue(sessionData);
+  const [primeCacheWithProfile, data] = useProfileFromCache(sessionData);
+
+  const result = resultFrom(sessionData, data);
+
+  const update = async (newSessionData: SessionData | null) => {
+    invariant(newSessionData, 'Session data must be defined.');
+
+    if (newSessionData.type === SessionType.WithProfile) {
+      const response = await primeCacheWithProfile(newSessionData.profileId);
+
+      if (response.error) {
+        return;
+      }
+
+      invariant(response.data, 'Cannot fetch profile data');
+    }
+  };
+
+  useEffect(() => {
+    if (result.loading) {
+      if (sessionData?.type === SessionType.WithProfile) {
+        void update(sessionData);
+        return;
+      }
+
+      return sessionDataVar.onNextChange(async function onNext(newSessionData) {
+        await update(newSessionData);
+      });
+    }
+    return;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // Handle suspense
+  if (args?.suspense) {
+    if (result.loading) {
+      // Do the same thing as the effect right above because the effect won't run
+      // when we suspend.
+
+      throw new Promise<void>((resolve) => {
+        sessionDataVar.onNextChange(async (newSessionData) => {
+          await update(newSessionData); // TODO serve????
```

Comment:
I see that throwing a promise is a react suspense thing, anything else you wanted to add with your TODO comment?

Reply:
no, I forgot to remove the TODO.